File: js/src/expiring.ts
import { alive } from "./alive.js";
import { effectiveStatus, loadRegistry, parseDate } from "./registry.js";
import type { AliveResult } from "./types.js";

export function listExpiring(
  opts: { withinDays?: number; provider?: string; today?: Date } = {},
): AliveResult[] {
  const withinDays = opts.withinDays ?? 30;
  const today = opts.today ?? new Date();
  const registry = loadRegistry();
  const results: AliveResult[] = [];

  for (const [modelId, entry] of Object.entries(registry.models)) {
    if (opts.provider && entry.provider !== opts.provider) continue;
    const status = effectiveStatus(entry, today);
    if (status !== "deprecated" && status !== "retired") continue;
    const retiredAt = parseDate(entry.retired_at);
    if (!retiredAt) continue;
    const ms = retiredAt.getTime() - today.getTime();
    const daysLeft = Math.floor(ms / (1000 * 60 * 60 * 24));
    if (status === "retired" || daysLeft < 0) continue;
    if (daysLeft <= withinDays) {
      results.push(alive(modelId, today));
    }
  }

  return results.sort(
    (a, b) => (a.days_until_retirement ?? 9999) - (b.days_until_retirement ?? 9999),
  );
}

File: modelalive/expiring.py
from __future__ import annotations

from datetime import date
from pathlib import Path

from modelalive.lifecycle import effective_status, parse_date
from modelalive.registry import load_registry
from modelalive.types import AliveResult
from modelalive.check import alive


def list_expiring(
    *,
    within_days: int = 30,
    provider: str | None = None,
    registry_path: str | Path | None = None,
    today: date | None = None,
) -> list[AliveResult]:
    current = today or date.today()
    registry = load_registry(registry_path)
    results: list[AliveResult] = []

    for model_id, entry in registry.get("models", {}).items():
        if provider and entry.get("provider") != provider:
            continue
        status = effective_status(entry, today=current)
        if status not in {"deprecated", "retired"}:
            continue
        retired_at = parse_date(entry.get("retired_at"))
        if retired_at is None:
            continue
        days_left = (retired_at - current).days
        if status == "retired" or days_left < 0:
            continue
        if days_left <= within_days:
            results.append(alive(model_id, registry_path=registry_path, today=current))

    return sorted(results, key=lambda r: r.days_until_retirement or 0)
